### PR TITLE
Compact classnames for subcomponent implementations, fixes #421.

### LINKED ISSUE #421: File name too long when nesting subcomponents
I'm going through the final steps to fully convert our app to Dagger 2, tying the hierarchy of subcomponents together. We have a deep hierarchy of sub components. This means the subcomponent names are too long:

```
error while writing DaggerDevRegisterAppComponent.DevLoggedInComponentImpl.DevRootActivityComponentImpl.SellerFlow_MobileComponentImpl.HomeScreen_MobileComponentImpl.TenderPath_ComponentImpl.AbstractGiftCardBalancePath_ComponentImpl.GiftCardBalanceInputScreen_ComponentImpl: 
./DaggerDevRegisterAppComponent$DevLoggedInComponentImpl$DevRootActivityComponentImpl$SellerFlow_MobileComponentImpl$HomeScreen_MobileComponentImpl$TenderPath_ComponentImpl$AbstractGiftCardBalancePath_ComponentImpl$GiftCardBalanceInputScreen_ComponentImpl.class (File name too long)
```

### PR DESCRIPTION
**Pull request by rjrjr**
Fix for https://github.com/google/dagger/issues/421.

Instead of including more and more levels of container names until we achieve
uniqueness, we try two strategies: use simple names everywhere, or else
include a uniquing prefix everywhere.

The prefix is built from the initials of a containing class for nested
subcomponent declarations, or the initials of containing packages for
top level. An integer may appended to the prefix if necessary.

### CODE REVIEW

**Comment by ronshapiro** [comment]
I think once we get names like `F2_SubImpl` or `tdmltmt_SubImpl`, I think we've lost all value of having recognizeable names, and something like `{simple_name}{number}Impl` is probably just as good.

If we defined subcomponents as static classes, all nested one level inside the generated component, we can keep the more clear class names, but at the cost of needing to maintain our own pointers back to the parent components. There are other bugs that people have filed that want that (I think @pyricau had some).

**Comment by rjrjr** [comment]
I tried simple names first, but they're just too long, we still had the same problem. WRT the readability of `F2_SubImpl`, test cases like that are not very realistic examples. In real code we see things like:

```
  @Override
  public SellerScope.PhoneComponent sellerScopePhone() {
    return new SS_PhoneComponentImpl();
  }

  @Override
  public SellerScope.TabletComponent sellerScopeTablet() {
    return new SS_TabletComponentImpl();
  }
```

and

```
  private final class SS_TabletComponentImpl implements SellerScope.TabletComponent {
```

In context I found it pretty readable -- I imagine that's why @gk5885 first suggested this approach. And such terms always appear next to something clickable for IDE navigation. 

I could take a stab at the static classes, but I'll need some handholding (and there's really something bothersome about reimplementing a language feature that way). It's your call, obviously, but I'd be happier landing this adequate solution.

**Comment by rjrjr** [comment]
@ronshapiro Ping. Waiting to hear if you have been overwhelmed by the share elegance of my argument to merge this as-is, or if you can give me (very detailed) pointers on how to tackle the static class approach.

**Comment by ronshapiro** [comment]
I think @gk5885 was more in favor of this then I was. The static class approach is likely to be a big undertaking, since there are a lot of parts that implicitly rely on the current structure. If once Greg chimes in we think it's worth considering, it probably makes more sense to discuss the static classes in a design doc format or GVC

**Comment by gk5885** [comment]
I like this strategy a lot -- there's really no ambiguity if the name doesn't collide and the source is likely to be as easy to navigate as it would be otherwise.  Ship it!

**Comment by ronshapiro** [comment]
@gk5885 do you want to do a codereview here and then let me know and I'll pull it in?

File: compiler/src/main/java/dagger/internal/codegen/ComponentWriter.java [hunk 1]
```diff
@@ -103,36 +108,89 @@ private UniqueSubcomponentNamesGenerator(BindingGraph graph) {
     private ImmutableBiMap<ComponentDescriptor, String> disambiguateConflictingSimpleNames(
         Collection<ComponentDescriptor> components) {
       Map<String, ComponentDescriptor> generatedSimpleNames = new LinkedHashMap<>();
-      // The ending condition is when there is a unique simple name generated for every element
-      // in components. The sizes should be equivalent (with one generated name per component).
-      for (int levels = 0; generatedSimpleNames.size() != components.size(); levels++) {
+
+      // Let's see if we can get away with using simpleName() everywhere.
+      for (ComponentDescriptor component : components) {
```
Comment (gk5885) [inline compiler/src/main/java/dagger/internal/codegen/ComponentWriter.java]:
```
components
    .stream()
    .map(componentNamers::get)
    .map(Namer::simpleName)
    .allMatch(new HashSet<>()::add)
```
(or something like that) will tell you if you have a unique collection

File: compiler/src/main/java/dagger/internal/codegen/ComponentWriter.java [hunk 2]
```diff
@@ -103,36 +108,89 @@ private UniqueSubcomponentNamesGenerator(BindingGraph graph) {
...
-            // 2+ subcomponent descriptors will have the same simple name
-            break;
-          }
+          Namer namer = componentNamers.get(component);
+          final String simpleName = namer.simpleName();
+          final String basePrefix = namer.uniquingPrefix();
+          String candidateName;
+          int disambiguate = 0;
+
+          do {
```
Comment (gk5885) [inline compiler/src/main/java/dagger/internal/codegen/ComponentWriter.java]: Can you use `dagger.internal.codegen.UniqueNameSet` for this?

**Comment by rjrjr** [comment]
Belated thanks for the review!

I'm working on this now. Do you like updates squashed?

**Comment by ronshapiro** [comment]
Don't worry about squashing. I'll take care of that when we merge
internally.

File: compiler/src/main/java/dagger/internal/codegen/ComponentWriter.java (hunk 1 above)
Comment (rjrjr) [inline compiler/src/main/java/dagger/internal/codegen/ComponentWriter.java]: Sorry, my stream fu is weak. If this passes, won't it require me to iterate a second time to actually build the map?

**Comment by ronshapiro** [comment]
@rjrjr @gk5885 is this ready to be submitted internally?

**Comment by rjrjr** [comment]
@ronshapiro I was away, back now. Looks like I need to rebase, and I need @gk5885 guidance on how to do the stream thing at line 113.

**Comment by rjrjr** [comment]
@ronshapiro This looks OK to merge to me if @gk5885 approves. I believe I've addressed all feedback except for the stream thing, I'm just not functional enough for that yet.

**Comment by rjrjr** [comment]
Man, getting serious flashbacks doing these `bazel` builds.

File: compiler/src/main/java/dagger/internal/codegen/ComponentWriter.java [hunk 3]
```diff
@@ -102,36 +108,77 @@ private UniqueSubcomponentNamesGenerator(BindingGraph graph) {
...
-              QUALIFIED_NAME_JOINER.join(
-                  pieces.subList(Math.max(0, pieces.size() - levels - 1), pieces.size()));
-          ComponentDescriptor conflict = generatedSimpleNames.put(simpleName, component);
-          if (conflict != null) {
-            // if the map previously contained an entry for the same simple name, stop early since
-            // 2+ subcomponent descriptors will have the same simple name
-            break;
-          }
+          Namer namer = componentNamers.get(component);
+          final String simpleName = namer.simpleName();
```
Comment (ronshapiro) [inline compiler/src/main/java/dagger/internal/codegen/ComponentWriter.java]: Can you remove `final` from this and the next line?

Comment (rjrjr) [inline compiler/src/main/java/dagger/internal/codegen/ComponentWriter.java]: Done, pushed.

**Comment by rjrjr** [comment]
@gk5885 ping

**Comment by ronshapiro** [comment]
This was actually submitted a few days ago in https://github.com/google/dagger/commit/ecba30a3fe3bf3e42c474ef5677f69f488569f53 but for some reason Github didn't auto-close it

**Comment by ronshapiro** [comment]
Oh, I didn't list this PR also in the Fixes list, thats why.

**Comment by rjrjr** [comment]
Wooo HOOOO! Thanks.

**Comment by ronshapiro** [comment]
Thanks for contributing it! Hope you can now remove your long lived fork!

File: compiler/src/main/java/dagger/internal/codegen/ComponentWriter.java [hunk 4]
```diff
@@ -102,36 +108,77 @@ private UniqueSubcomponentNamesGenerator(BindingGraph graph) {
...
-            break;
-          }
+          Namer namer = componentNamers.get(component);
+          String simpleName = namer.simpleName();
+          String basePrefix = namer.uniquingPrefix();
+          generatedSimpleNames.put(format("%s_%s", nameSet.getUniqueName(basePrefix), simpleName),
+              component);
         }
       }
       return ImmutableBiMap.copyOf(generatedSimpleNames).inverse();
```
Comment (naturalwarren) [inline compiler/src/main/java/dagger/internal/codegen/ComponentWriter.java]: Won't this actually result in longer names when simple names are used now that we're not splitting the qualified name?

File: compiler/src/main/java/dagger/internal/codegen/ComponentWriter.java [hunk 5]
```diff
@@ -102,36 +108,77 @@ private UniqueSubcomponentNamesGenerator(BindingGraph graph) {
...
+
+      if (generatedSimpleNames.size() != components.size()) {
+        // Simple approach didn't work out, let's use more complicated names.
+        // We keep them small to fix https://github.com/google/dagger/issues/421.
         generatedSimpleNames.clear();
+        UniqueNameSet nameSet = new UniqueNameSet();
         for (ComponentDescriptor component : components) {
-          List<String> pieces = componentQualifiedNamePieces.get(component);
-          String simpleName =
-              QUALIFIED_NAME_JOINER.join(
```
Comment (naturalwarren) [inline compiler/src/main/java/dagger/internal/codegen/ComponentWriter.java]: Won't this actually result in longer names when simple names are used now that we're not splitting the qualified name?

File: compiler/src/main/java/dagger/internal/codegen/ComponentWriter.java [hunk 6]
```diff
@@ -102,36 +108,77 @@ private UniqueSubcomponentNamesGenerator(BindingGraph graph) {
...
 
-    private static ImmutableListMultimap<ComponentDescriptor, String> qualifiedNames(
+    private static ImmutableMap<ComponentDescriptor, Namer> qualifiedNames(
         Iterable<ComponentDescriptor> componentDescriptors) {
-      ImmutableListMultimap.Builder<ComponentDescriptor, String> builder =
-          ImmutableListMultimap.builder();
+      ImmutableMap.Builder<ComponentDescriptor, Namer> builder = ImmutableMap.builder();
       for (ComponentDescriptor component : componentDescriptors) {
-        Name qualifiedName = component.componentDefinitionType().getQualifiedName();
-        builder.putAll(component, QUALIFIED_NAME_SPLITTER.split(qualifiedName));
```
Comment (naturalwarren) [inline compiler/src/main/java/dagger/internal/codegen/ComponentWriter.java]: Won't this actually result in longer names when simple names are used now that we're not splitting the qualified name?

**Comment by ronshapiro** [comment]
Someone internally just ran into this, where each component had a different simple name and thus this renaming strategy didn't come into effect. We may need to implement another pass that checks whether or not the concatenated name of a nested subcomponent is too long, and therefore we should shorten names until we get under the threshold

**Pull request closed without merging**